### ISSUE TITLE
Detector - added iou_thresh=0.213 parameter to use more than 1 anchor per truth

### ISSUE THREAD

**Issue by AlexeyAB** [AlexeyAB/darknet#4451]
Now you can use
```
[yolo]
iou_thresh=0.213
```
So will be used more than 1 anchors per object (truth), not only the best anchor but also all anchors with `IoU > 0.213`

Related to: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-557835331

Commits: 
* https://github.com/AlexeyAB/darknet/commit/e7e85b358cb0531f7154fdd68306c4c4dc96b5d2
* https://github.com/AlexeyAB/darknet/commit/0cf4c16c9fca93502333fef383e1287b38ca38f8

**Comment by AlexeyAB** [comment]
@glenn-jocher Hi,

What mAP gain did you get by using each of your improvements?
1. `iou_thresh=0.213` (more than 1 anchor per truth) - current issue https://github.com/AlexeyAB/darknet/issues/4451
2. `Mosaic data augmentation` - https://github.com/AlexeyAB/darknet/issues/4264
3. `[yolo] iou_normalizer=0.07 cls_normalizer=1.0 iou_loss=giou` (iou 3.31, cls 42.4, obj 52.0) https://github.com/AlexeyAB/darknet/issues/4430

**Comment by LukeAI** [comment]
This gave me over 1mAP boost

CSresnext50-spp-pan-scale
![csresnextleaky](https://user-images.githubusercontent.com/43993778/70320928-d7b06000-181d-11ea-8f35-353cf21728dd.png)

CSresnext50-spp-pan-scale-iou
![chart](https://user-images.githubusercontent.com/43993778/70320952-e860d600-181d-11ea-8d8a-7aab023f581b.png)

**Comment by AlexeyAB** [comment]
@LukeAI 

Also try: `[yolo] iou_normalizer=0.07 cls_normalizer=1.0 iou_loss=giou` (iou 3.31, cls 42.4, obj 52.0)

**Comment by glenn-jocher** [comment]
> @glenn-jocher Hi,
> 
> What mAP gain did you get by using each of your improvements?
> 
> 1. `iou_thresh=0.213` (more than 1 anchor per truth) - current issue #4451
> 2. `Mosaic data augmentation` - #4264
> 3. `[yolo] iou_normalizer=0.07 cls_normalizer=1.0 iou_loss=giou` (iou 3.31, cls 42.4, obj 52.0) #4430

@AlexeyAB Unfortunately I don't have exact ablation test results to give you. The iou_thresh, if set to 0.0, will return very poor mAPs, because every anchor is trying to fit every target. Mosaic was extremely important for `preventing validation set overfitting` during training, which allows us to train longer (i.e. to 500k iterations) and increase mAP. 

The normalization part seems to be integral to every neural network, with a good rule of thumb being that **multicomponent losses (like we use in object detection) should contribute proportionally to their importance to the metric** (mAP). Hyperparameter evolution is an effective but extremely slow way to do this. I'm thinking a good shortcut might be to train for one epoch with default normalizers in place, then update the 3 normalizers (box, obj, cls) to bring all of the losses back into balance, targeting a median value of say 5 each at the end of the first epoch. Logic like this might help the community in general train more effective models more easily, as I've noticed that COCO hyperparameters are not always optimal on 3rd party datasets.

**Comment by AlexeyAB** [comment]
@glenn-jocher 
Do you mean that we should try to update the 3 normalizers (box, obj, cls) for each epoch, so each of 3 losses will be equal?

I added such ToDo. https://github.com/AlexeyAB/darknet/issues/4463

**Comment by glenn-jocher** [comment]
@AlexeyAB you know I tried that once, but it didn't work very well. Maybe they should naturally diverge as training progresses, not sure.

In any case probably balancing them once before actual training begins is probably a good place to start, or maybe balancing them after just the first or second epochs maybe.
